pytorch_lightning/callbacks/device_stats_monitor.py

```diff
@@ -0,0 +1,62 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Device Stats Monitor
+=================
+Monitors and logs device stats during training.
+"""
+from typing import Any, Optional
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+
+
+class DeviceStatsMonitor(Callback):
+    r"""
+    Automatically monitors and logs device stats during training stage. ``DeviceStatsMonitor``
```

Comment:
do we also want to report stats during validation, test, and predict phases?

Reply:
sure, I can add that. Currently GPUStatsMonitor and XLAStatsMonitor are just for training stage, which is the model I followed.

Reply:
for this one, how do you envision adding this? It seems `on_batch_start` and `on_batch_end` don't apply to valid/test/predict (#9727) @ananthsub

Reply:
you could also do it with `on_{train/val/test/predict}_batch_start` hooks, and share the same implementation across all of them.

with the current implementation, this will log the same metric key across training/validation/test/predict without an indication as to which phase they're occurring in. this could be misleading when viewing the logged data as training will have a different device profile compared to val/test/predict because of the extra comms, backwards & optimizer steps involved.

in order to merge this, and unblock the deprecation of `log_gpu_memory` in the Trainer constructor, what do you think of proceeding with just logging during training for now, and in a separate PR extend support for all of the running stages. to do so, we'd likely need to prefix the keys with the running stage (and maybe the rank) and a way to group these into a single section in the logger e.g. `Devices/{running_stage}/{metric_key}`. i apologize for the thrash here

Reply:
sounds good makes sense. updated PR